alacritty/src/renderer/rects.rs

```diff
@@ -190,3 +195,205 @@ impl RenderLines {
         }
     }
 }
+
+/// Shader sources for rect rendering program.
+pub static RECT_SHADER_F_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/res/rect.f.glsl");
+pub static RECT_SHADER_V_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/res/rect.v.glsl");
+static RECT_SHADER_F: &str = include_str!("../../res/rect.f.glsl");
+static RECT_SHADER_V: &str = include_str!("../../res/rect.v.glsl");
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Rgba {
+    r: u8,
+    g: u8,
+    b: u8,
+    a: u8,
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Vertex {
+    // Normalized screen coordinates.
+    x: f32,
+    y: f32,
+
+    // Color.
+    color: Rgba,
+}
+
+#[derive(Debug)]
+pub struct RectRenderer {
+    // GL buffer objects.
+    vao: GLuint,
+    vbo: GLuint,
+
+    program: RectShaderProgram,
+
+    vertices: Vec<Vertex>,
+}
+
+impl RectRenderer {
+    /// Update program when doing live-shader-reload.
+    pub fn set_program(&mut self, program: RectShaderProgram) {
+        self.program = program;
+    }
+
+    pub fn new() -> Result<Self, renderer::Error> {
+        let mut vao: GLuint = 0;
+        let mut vbo: GLuint = 0;
+        let program = RectShaderProgram::new()?;
+
+        unsafe {
+            // Allocate buffers.
+            gl::GenVertexArrays(1, &mut vao);
+            gl::GenBuffers(1, &mut vbo);
+
+            gl::BindVertexArray(vao);
+
+            // VBO binding is not part of VAO itself, but VBO binding is stored in attributes.
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+
+            // Position.
+            gl::VertexAttribPointer(
+                0,                          // Attribute location.
+                2,                          // Component count: 2. x and y.
+                gl::FLOAT,                  // Attribute type: f32
+                gl::FALSE,                  // Not normalized.
+                size_of::<Vertex>() as i32, // Stride.
+                std::ptr::null(),           // Offset of Vertex::x is zero.
+            );
+            gl::EnableVertexAttribArray(0);
+
+            // Color.
+            gl::VertexAttribPointer(
+                1,                          // Attribute location.
+                4,                          // Component count: 4. r, g, b, a.
+                gl::UNSIGNED_BYTE,          // Attribute type: u8.
+                gl::TRUE,                   // Normalized: 0..255 is mapped to 0..1.
+                size_of::<Vertex>() as i32, // Stride.
+                8 as *const _,              // Offset of Vertex::color.
```

Comment:
Could you apply the logic with `size` we have inside a macro in quad renderer? So it's a bit more clear what it's going on than `8`?